File: game/src/Paddle.lua
local GameObject = require "src.GameObject"
local Object = require "lib.Object"
local Vector2 = require "lib.Vector2"

local DEFAULT_PADDLE_SIZE = {
    x = 120,
    y = 10,
}

local DEFAULT_ROUNDING_SEGMENTS = {
    x = DEFAULT_PADDLE_SIZE.x * 0.2,
    y = DEFAULT_PADDLE_SIZE.y * 0.2,
}

---@class Paddle
---@overload fun(ctx: Game.Context, state?: Paddle.State): self

---@class Paddle: Game.Object, Paddle.State
local Paddle = GameObject:extend()

---@param ctx Game.Context
---@param state? Paddle.State
function Paddle:init(ctx, state)
    self:initState(state, function()
        ---@class Paddle.State
        local state = {
            ---Angle from centre to mouse
            ---@type number
            angle = 0.0,

            ---Max speed
            ---@type number
            speed = 500.0,

            ---Distance relative to mouse
            ---@type number
            distance = 0.0,

            ---Current speed
            ---@type number
            velocity = 0.0,

            ---Position of the sprite
            position = Vector2(),

            ---Center point of the sprite
            centre = Vector2(),

            ---Size of the sprite
            size = Vector2(DEFAULT_PADDLE_SIZE),

            ---Corner rounding values
            rounding = Vector2(DEFAULT_ROUNDING_SEGMENTS),
        }
        return state
    end)
    self.position.y = ctx.window.size.y - 2 * self.size.y
end

---Used for better readability.<br>
---If ever multiple paddles should interact with each other, this must not be used in the corresponding functions.<br>
---In theory, since update functions are executed sequentially, there should be no race condition.
---However, if asynchronicity or multithreading is ever introduced into the game logic, this pattern has to disappear first.
---@type Paddle.State
local state

---@param ctx Game.Context
function Paddle:update(ctx)
    if ctx.keyboard.space:getState() == "down" then
        -- print(function()
        --     return ("\n"):rep(50) .. "<Space> held -> Paddle stopped"
        -- end)
        return
    elseif ctx.keyboard.space:getState() == "pressed" then
        -- print(function()
        --     return ("\n"):rep(50)
        -- end)
    end

    local xMouse, yMouse = ctx.mouse:getPosition()
    state = self.state

    state.distance = math.abs(xMouse - state.centre.x)
    if state.distance < 5 then
        state.velocity = 0
    elseif state.distance < state.size.x * 2 then
        state.velocity = math.min(state.speed, state.speed * (state.distance / 20))
    end

    state.angle = math.atan2(yMouse - state.centre.y, xMouse - state.centre.x)
    state.position.x = state.position.x + state.velocity * math.cos(state.angle) * ctx.deltaTime
    state.position.x = math.floor(state.position.x)

    if state.position.x < 0 then
        state.position.x = 0
    elseif state.position.x + state.size.x > ctx.window.size.x then
        state.position.x = ctx.window.size.x - state.size.x
    end

    state.centre.x = state.position.x + state.size.x / 2
    state.centre.y = state.position.y + state.size.y / 2
end

---@param ctx Game.Context
function Paddle:draw(ctx)
    state = self.state

    love.graphics.rectangle(
        "fill",
        state.position.x,
        state.position.y,
        state.size.x,
        state.size.y,
        state.rounding.x,
        state.rounding.y
    )
end

return Paddle

File: game/src/Paddle.ts
import { InputState, Keyboard, Mouse, Window } from 'context'
import { Rectangle } from 'classes/Shapes'
import { Vector2 } from 'classes/Vector'

const WIDTH = 120
const HEIGHT = 10

declare global {
  interface CustomHandlers {
    ['paddle/update']: () => void
  }
}

export class Paddle extends Rectangle {
  private distance: number = math.huge
  private velocity: number = 0
  private angle: number = 0

  public readonly centre: Vector2
  public readonly speed: number = 500

  constructor() {
    super(WIDTH, HEIGHT, new Vector2(0, Window.height - HEIGHT * 3))

    this.centre = new Vector2(
      this.origin.x + this.width / 2,
      this.origin.y + this.height / 2
    )
  }

  updateDistance() {
    this.distance = math.abs(Mouse.position.x - this.centre.x)
  }

  updateAngle() {
    this.angle = this.centre.angle(Mouse.position)
  }

  updatePosition(dt: number) {
    this.origin.x = this.origin.x + math.cos(this.angle) * this.speed * dt // TODO: Add proper easing
    this.origin.x = math.clamp(this.origin.x, 0, Window.width - this.width)
    this.origin.x = math.round(this.origin.x)
    this.centre.x = this.origin.x + this.width / 2
    this.centre.y = this.origin.y + this.height / 2
  }

  update(dt: number) {
    this.updateDistance()
    if (this.distance <= 2) return
    this.updateAngle()
    if (Keyboard.get('space') === InputState.DOWN) return
    this.updatePosition(dt)
  }

  draw() {
    love.graphics.rectangle(
      'line',
      this.origin.x,
      this.origin.y,
      this.width,
      this.height
    )
  }
}
